### DISCUSSION TITLE
Respect compiler directives

### DISCUSSION THREAD

**Discussion by cjnevin-foxtel** [gohanlon/swift-memberwise-init-macro#45] [Ideas]
This one has been irking me for a while now and I haven't had a chance to investigate and add it myself.

Basically, I'd like to be able to define properties only for particular platforms or environments.

# OS Examples
```swift
@MemberwiseInit(.public)
public struct Dependencies {
   public let common: Common
   #if os(iOS)
   public let mobile: Mobile
   #elseif os(macOS)
   public let desktop: Desktop
   #elseif os(tvOS)
   public let tv: TV
   #endif
}

// iOS
Dependencies.init(common: Common, mobile: Mobile)
// macOS
Dependencies.init(common: Common, desktop: Desktop)
// tvOS
Dependencies.init(common: Common, tv: TV)
```

Depending on platform we can then define different variables in the same struct.

Currently I'd have to define 3 structs that are almost identical to generate a slightly different initializer for each or provide a struct named the same but containing different properties.

# Debug vs Release

We can also do the same thing with #if DEBUG to only provide variables if we are debugging or provide different default values if we are debugging that may log their output for example.

**Comment by gohanlon** [comment]
Unfortunately, @MemberwiseInit can't support this. Swift Macros do receive the raw syntax of `#if` declarations but can't evaluate conditions like `os(iOS)`. Supporting this would likely require significant changes to Swift Macros' design.
